integration_tests/src/main/python/data_gen.py

```diff
@@ -208,6 +209,15 @@ def __init__(self, nullable=True, min_val = INT_MIN, max_val = INT_MAX,
     def start(self, rand):
         self._start(rand, lambda : rand.randint(self._min_val, self._max_val))
 
+
+class DecimalGen(DataGen):
+    """Generate Decimals, with some built in corner cases."""
+    def __init__(self, nullable=True):
+        super().__init__(DecimalType(7,3), nullable=nullable)
+
+    def start(self, rand):
+        self._start(rand, lambda : Decimal(str(round(random.uniform(2000, 4000), 2))))
```

Comment:
I really only see one good choice when generating random data for Decimal.  We need to use `sre_yield` like with the StringGen.  We need to create a pattern that will work.  Essentially we need to have a pattern like

```
"-?[0-9]{1," + str(precision-scale) + "}.[0-9]{0," + str(scale) + "}"
```

Note that by default Spark does not support a negative scale (That is controlled by `park.sql.legacy.allowNegativeScaleOfDecimal`) but we do need to support it for testing.  So the above pattern only works for positive scale.  For negative scale we would need something more like.

```
"-?[0-9]{1,"+str(precision) + "}e" + str(-scale)
```

Note I have not tested these, this is just off the top of my head.

Reply:
Your PR covered this.